compiler/rustc_symbol_mangling/src/typeid/typeid_itanium_cxx_abi.rs

```diff
@@ -0,0 +1,404 @@
+        ty::Tuple(_) if ty.is_unit() => {
+            typeid.push('v');
+        }
+
+        // Sequence types
+        ty::Tuple(tys) => {
+            // u5tupleI<element-type1..element-typeN>E as vendor extended type
+            let mut s = String::from("u5tupleI");
+            for ty in tys.iter() {
+                s.push_str(&encode_ty(tcx, ty, dict, encode_c_types));
+            }
+            s.push('E');
+            compress(dict, ty, &mut s);
+            typeid.push_str(&s);
+        }
+
+        ty::Array(ty0, len) => {
+            // u5arrayI<array-length><element-type>E as vendor extended type
+            let mut s = String::from("u5arrayI");
+            s.push_str(
+                format!("{}", len.val().try_to_scalar().unwrap().to_i64().unwrap()).as_ref(),
+            );
+            s.push_str(&encode_ty(tcx, *ty0, dict, encode_c_types));
+            s.push('E');
+            compress(dict, ty, &mut s);
+            typeid.push_str(&s);
+        }
+
+        ty::Slice(ty0) => {
+            // u5sliceI<element-type>E as vendor extended type
+            let mut s = String::from("u5sliceI");
+            s.push_str(&encode_ty(tcx, *ty0, dict, encode_c_types));
+            s.push('E');
+            compress(dict, ty, &mut s);
+            typeid.push_str(&s);
+        }
+
+        // User-defined types
+        // FIXME(rcvalle): Identify C char and integer type uses and encode them with their
+        // respective builtin type encoding as specified by the Itanium C++ ABI if encode_c_types
+        // is true.
+        ty::Adt(adt_def, ..) => {
+            // S<length><name>
+            let mut s = String::new();
+            let def_id = adt_def.0.did;
+            let name = format!("{}", tcx.opt_item_name(def_id).unwrap());
+            let crate_name = format!("{}", tcx.crate_name(def_id.krate));
+            if name == "c_void"
+                && (crate_name == "core" || crate_name == "std" || crate_name == "libc")
+            {
+                s.push('v');
+            } else {
+                s.push_str(&format!("{}", name.len()));
+                s.push_str(&name);
+                compress(dict, ty, &mut s);
+            }
+            typeid.push_str(&s);
+        }
+
+        ty::Foreign(def_id) => {
+            // S<length><name>
+            let mut s = String::new();
+            let name = format!("{}", tcx.opt_item_name(*def_id).unwrap());
+            s.push_str(&format!("{}", name.len()));
+            s.push_str(&name);
+            compress(dict, ty, &mut s);
+            typeid.push_str(&s);
+        }
+
+        // Function types
+        ty::FnDef(def_id, ..) => {
+            // u2fnI<length><name>E
+            let mut s = String::from("u2fn");
+            let name = format!("{}", tcx.opt_item_name(*def_id).unwrap());
+            s.push_str(&format!("{}", name.len()));
+            s.push_str(&name);
+            s.push('E');
```

Comment:
The opening `I` is missing.

(I was looking at the codegen test case filtered through c++filt / llvm-cxxfilt, but it seems that template arguments are currently unsupported in vendor extended types, so types using those production don't demangle either way).

Reply:
Oops! Fixed.